### PR TITLE
Implement per-update cache for roles and collections

### PR DESCRIPTION
**Pull request by AlanCoding**
##### SUMMARY
This is a direct continuation of https://github.com/ansible/awx/pull/7194

Connect https://github.com/ansible/awx/issues/5518

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - API

##### AWX VERSION
```
13.0.0
```

##### ADDITIONAL INFORMATION
I have to be really picky about what to refactor here.

See #5429, that has a lot of background for some of the patterns here. I think `make_local_copy` should be a method on the project model. I'm trying to allow for that in the future, but keep changes here minimal.

That PR also may reflect some shortcomings of this which still remain. For SCM inventory, roles might not be getting installed in some cases, and that still needs some investigation.

### CODE REVIEW

File: awx/main/tasks.py
Comment (wenottingham) [inline awx/main/tasks.py:1982]: ?

Comment (AlanCoding) [reply awx/main/tasks.py:1982]: ? as in, "why did remove the job dir here" or ? as in, "why did use `@property` when I could have make it an actual property?"

For the latter question, the answer is CTiT, I don't want it to be looking up the setting on load of the module.

The former question is much bigger. I removed the `synchronize` module from the playbook. By doing this, the duties of the project update playbook become much more static, because given a certain project, it always downloads to the same "stage" folder for the roles/collections. Thinking there relates to https://github.com/ansible/awx/issues/7625, I want to have a project update interface that works for that, while also working for caching roles / collections. The project cache folder is already a subfolder of the PROJECTS_ROOT folder, so it no longer needs to be added here.

I'm going to be writing docs on this. The overall function will be much easier to explain with these changes.

File: awx/playbooks/project_update.yml
Comment (AlanCoding) [inline awx/playbooks/project_update.yml]: My intent here is that `project_path` will be removed, replaced with `{{projects_root}}/{{local_path}}` in the playbook. After that, the playbook will no longer document this variable, or have it passed in extra_vars.

I just didn't want to do that in this diff, but I could be convinced otherwise.

**Comment by AlanCoding** [comment]
I do have a few other corner case concerns:

 - [x] the project update `pre_run_hook` (with lock held) should assure that the "stage" directory doesn't exist. It never should, so if it does, it should log warnings
 - [x] we _might_ need to clear the cache more aggressively for the branch override cases, because it could leave an unlimited number of caches around. This might be workable by clearing all but `last_job_id` in both cases...

I'm still trying to challenge myself to come up with more potential issues. I feel much better about SCM inventory syncs after the last few commits, I expect it will probably work right, but might give erroneous warnings, not fully utilize the cache, and still be rough around the edges. I would hate to wind up effectively re-doing #5429, but I think it may come to that.

**Comment by AlanCoding** [comment]
My only remaining issue with the code as it stands is that the logic for needs roles/collections is much more complicated than it needs to be to cover cases we would actually hit. I may or may not make another change for that.

**Comment by AlanCoding** [comment]
I think that this is cleaner, but it should make any functional difference, and I won't put it in this branch until I'm sure it will not break things.

```diff
diff --git a/awx/main/tasks.py b/awx/main/tasks.py
index 6bad842bce..f9ed09ab28 100644
--- a/awx/main/tasks.py
+++ b/awx/main/tasks.py
@@ -1872,48 +1872,34 @@ class RunJob(BaseTask):
         elif not os.path.exists(project_path):
             logger.debug('Performing fresh clone of {} on this instance.'.format(job.project))
             sync_needs = all_sync_needs
-        elif not job.project.scm_revision:
-            logger.debug('Revision not known for {}, will sync with remote'.format(job.project))
-            sync_needs = all_sync_needs
-        elif job.project.scm_type == 'git':
+        elif job.project.scm_type == 'git' and job.project.scm_revision and (not branch_override):
             git_repo = git.Repo(project_path)
             try:
-                desired_revision = job.project.scm_revision
-                if branch_override:
-                    desired_revision = job.scm_branch  # could be commit or not, but will try as commit
-                current_revision = git_repo.head.commit.hexsha
-                if desired_revision == current_revision:
-                    job_revision = desired_revision
+                if job_revision == git_repo.head.commit.hexsha:
                     logger.debug('Skipping project sync for {} because commit is locally available'.format(job.log_format))
+                    cache_id = str(job.project.last_job_id)
+                    has_cache = os.path.exists(os.path.join(job.project.get_cache_path(), cache_id))
+                    if has_cache:
+                        logger.debug('Using cached roles or collections for {}'.format(job.log_format))
+                    else:
+                        # see if we need a sync because of presence of roles
+                        roles_exist = os.path.exists(os.path.join(project_path, 'roles', 'requirements.yml'))
+                        if roles_exist:
+                            logger.debug('Running project sync for {} for role requirements.'.format(job.log_format))
+                            sync_needs.append('install_roles')
+
+                        collections_exist = os.path.exists(os.path.join(project_path, 'collections', 'requirements.yml'))
+                        if collections_exist:
+                            logger.debug('Running project sync for {} for collection requirements.'.format(job.log_format))
+                            sync_needs.append('install_collections')
                 else:
                     sync_needs = all_sync_needs
             except (ValueError, BadGitName):
                 logger.debug('Needed commit for {} not in local source tree, will sync with remote'.format(job.log_format))
                 sync_needs = all_sync_needs
         else:
+            logger.debug('Project not available locally {}, will sync with remote'.format(job.project))
             sync_needs = all_sync_needs
-        # Galaxy requirements are not supported for manual projects
-        if not sync_needs and job.project.scm_type:
-            has_cache = False
-            if not branch_override:
-                cache_id = str(job.project.last_job_id)
-                has_cache = os.path.exists(os.path.join(job.project.get_cache_path(), cache_id))
-            # see if we need a sync because of presence of roles
-            galaxy_req_path = os.path.join(project_path, 'roles', 'requirements.yml')
-            roles_exist = os.path.exists(galaxy_req_path)
-            if roles_exist and not has_cache:
-                logger.debug('Running project sync for {} because of galaxy role requirements.'.format(job.log_format))
-                sync_needs.append('install_roles')
-            elif roles_exist:
-                logger.debug('Using cached roles for {}'.format(job.log_format))
-
-            galaxy_collections_req_path = os.path.join(project_path, 'collections', 'requirements.yml')
-            collections_exist = os.path.exists(galaxy_collections_req_path)
-            if collections_exist and not has_cache:
-                logger.debug('Running project sync for {} because of galaxy collections requirements.'.format(job.log_format))
-                sync_needs.append('install_collections')
-            elif collections_exist:
-                logger.debug('Using cached collections for {}'.format(job.log_format))
 
         if sync_needs:
             pu_ig = job.instance_group
@@ -2423,7 +2409,7 @@ class RunProjectUpdate(BaseTask):
                 instance.save(update_fields=['scm_revision'])
             # Roles and collection folders copy - both to durable cache and job dir
             stage_path = os.path.join(instance.get_cache_path(), 'stage')
-            if instance.branch_override:
+            if instance.branch_override or instance.job_type == 'check':
                 # We actually do not want to cache these cases, so we just set an id
                 # which we are sure that nothing will reference for its cache
                 cache_id = str(instance.id)
```

**Comment by AlanCoding** [comment]
Did another bit of a pivot, some major items changed:

 - [x] always create the stage dir in pre_run_hook
 - [x] always sync both roles and collections if either
        branch_override, or cache dir does not exist
 - [x] (need either way) only move stage to cache if successful
 - [x] change creation of stage to mkdirs
 - [x] do not populate cache if did not run galaxy tags

File: awx/main/tasks.py
Comment (wenottingham) [reply awx/main/tasks.py:1982]: Thanks, it was the former question, it wasn't obvious to me from looking at it.

File: awx/main/tasks.py
Comment (AlanCoding) [inline awx/main/tasks.py:2315]: You could also test `os.path.exists`, but I don't think it should matter at all. I really don't expect this check is needed given the other changes. However, I want to keep the check as a defensive measure. The contract may change in the future such that the stage dir is not always populated. Or error conditions could cause that...

File: awx/main/models/projects.py
Comment (AlanCoding) [inline awx/main/models/projects.py:623]: Let me get this out of the way:

With role/collection caching in place, it no longer makes _any_ sense to have 2 separate tags for these things. It should just be 1 tag, say "install_content". If you follow all the code here, you will see that one of them is never used without the other.

I am not consolidating those tags right now, because `job_tags` are a primary part of the model and something that the user sees. In the effort to maintain behavior consistency I'm not changing it now, but after this PR, we could/should make that change for `devel` branch.

File: awx/main/tasks.py
Comment (AlanCoding) [inline awx/main/tasks.py]: I'm still on the fence about the choices made here.

Background on change - here we are caching _content_, roles & collections, and marking that cache with an identifier. SCM inventory cannot use roles, only collections. But... an SCM inventory could trigger a local cache update, which later gets used by a job. So yes, we do need to update roles here, just to keep the cache mechanics simple.

However, I'm still not detecting whether the cache exists or not. That would be getting into what #5429 was trying to do. This means that the cache would be blown away and both roles & collections re-downloaded for every SCM inventory update. This is gory because right now it would leave a warning log:

https://github.com/ansible/awx/pull/7643/files#diff-9d4ea1dd908b35fb92eaede4bd10bb46R2409

So options are:
1. downgrade the `warn` log to `debug` or `info`, accept that we're not solving the caching problem for SCM inv
2. check if cache exists here, like in the `RunJob` pre_run_hook

This is a pained decision for me right now. I don't want to creep scope further, but I want this to make sense.


Since the SCM inventory obliged to download _roles_, I'm leaning toward option (2), because it would be seen as a regression for it to start doing 20 GB downloads in all these cases where the project does not even have collection requirements.

File: awx/main/tests/functional/test_tasks.py
Comment (AlanCoding) [inline awx/main/tests/functional/test_tasks.py:33]: We can't do `objects.create()` because that won't set `local_path`. It has to be created via the conical means of creating it _from_ the project.

File: awx/main/tests/unit/test_tasks.py
Comment (AlanCoding) [inline awx/main/tests/unit/test_tasks.py:66]: Again, `local_path` _has_ to be set for all non-manual projects. This is an oddball case because it's an object not even saved to the database, so our normal means of setting it won't work.

**Comment by AlanCoding** [comment]
I filed a related issue at https://github.com/ansible/awx/issues/7665

I've been diving into the `local_cache` problem because `local_path = os.path.basename(self.local_path)` causes me great anxiety. In two seconds flat, I'm sure the user would input `foo/`, which will give empty string, and potentially damage the integrity of the projects _root_ folder.

For the purposes of this PR, I have satisfied myself that it is automatically set by the API. I don't think there is any legacy data that might cause an issue, since most of this logic has been around for a very long time.

**Comment by AlanCoding** [comment]
@tvo318 I added some docs here, but are very technical.

For the user's experience, @gamuniz noted that this detail is something we should call out - my attempted wording:

> if the user hits the "update" button for a project, then the update will re-download roles even if the revision of the project did not change.

You might also add "otherwise, jobs will use roles cached from the last update", or something along those lines.

*everything also applies to collections. We might start saying content=(roles and collections), but I defer to your judgement there.

**Comment by AlanCoding** [comment]
@chrismeyersfsu is there any reason to keep these lines around?

https://github.com/ansible/awx/blob/b36c5d920993a8b06ea2e3f38fcccac87a13bc15/awx/main/tasks.py#L2036-L2040

I don't think they do what they claim to do, but I don't really know what's up with them. My preference would be to delete them.

**Comment by AlanCoding** [comment]
Things that we could do later, targeting `devel`:

 - ~do not have project syncs call `make_local_copy`, do in job/inventory task~
   ~do not pass job_private_data_dir to project update at all, it is none of its business~
   (no solution for non-git, I take it back)
 - consolidate logic to check for sync needs in job/inventory tasks
 - combine install_roles and install_collections tags into single tag
 - remove `project_path` from project_update.yml playbook variables, just use its components
 - copy instead of doing a shallow git clone #7367
 - simplify logic for project `local_path` setting

After this merges, I should probably refresh these and file an issue.

**Comment by AlanCoding** [comment]
I've thought about https://github.com/ansible/awx/pull/7643#issuecomment-661318048 more... no good can come from those lines of code. It was put there back when we assumed we were passing that directory to Galaxy CLI... we're not. Now we copy the folders from the cache. It gets worse. Sometimes we do hit this code path before a job, but other times we don't, so sometimes we're merging directories with different filesystem permissions. It's nothing but a liability. It must be destroyed.

**Comment by AlanCoding** [comment]
Tinkering with doing the things I mentioned in https://github.com/ansible/awx/pull/7643#issuecomment-661318780, it would look like this (have not ran any tests):

https://github.com/AlanCoding/awx/commit/02cda85d4015d3c49a25b75f0fb6572e5f7fab92

This convinces me that _now_ is not the right time to do these things. The most major expected gains are still not really possible because we support branch override for non-git types. That is something we could consider. Consolidating the sync needs code just makes it more obvious that we need to consolidate the code for _creating and running the sync_.

File: awx/main/tasks.py
Comment (AlanCoding) [inline awx/main/tasks.py]: whoops! missing the `.format`

File: awx/main/tasks.py
Comment (AlanCoding) [inline awx/main/tasks.py:2298]: Explanation: this is changed due to introduction of new property `branch_override`, and it's also true that `branch_override=True` implies `instance.job_type == 'run'` so I removed that from the condition.

File: awx/main/models/projects.py
Comment (shanemcd) [inline awx/main/models/projects.py:199]: Should we put a `logger.warn` in an `else` here?

Comment (ryanpetrello) [reply awx/main/models/projects.py:199]: Or perhaps a `.debug()`, `warn` might be too noisy.

File: awx/main/models/projects.py
Comment (chrismeyersfsu) [inline awx/main/models/projects.py:428]: This will never be a project `sync` job right?

File: awx/main/tasks.py
Comment (wenottingham) [inline awx/main/tasks.py:2320]: I initially added this out of an idea to do these deletes in the background outside of the lock.

I could never get that to work, though. (Just commenting for posterity)

File: awx/main/tasks.py
Comment (ryanpetrello) [inline awx/main/tasks.py:1869]: Why not just use `job.branch_override` here?

Comment (ryanpetrello) [reply awx/main/tasks.py:1869]: Oh, ignore me, this is a `Job`, not a `ProjectUpdate`.

File: awx/main/models/projects.py
Comment (AlanCoding) [reply awx/main/models/projects.py:199]: This is a mess. The only reason it's used is because this is the status-quo:

https://github.com/ansible/awx/blob/8b97e3f8bb168f77045ffb2f15cfffc8e77fa005/awx/main/models/projects.py#L191

If `local_path` is ever `None`, it will error in almost any case. We don't handle the null case. What would it even mean if you did get that? That would happen if `local_path` is `foo/`, but we auto-generate this field, and I can't figure out anyway the auto-generation would produce that. _Maybe_ manual projects could accept this kind of value. If it did, this wouldn't be right. You would prefer to `rstrip` the slash instead.

It looks like the code is, for some reason, expecting to be given an absolute path. I don't know why.

File: awx/main/models/projects.py
Comment (AlanCoding) [reply awx/main/models/projects.py:428]: You're looking at the `Project` model definition. It is also defined for `ProjectUpdate` down below. If it's a normal sync, then it return's the project's last job (for core functionality). If it's a branch override case, we agreed to give up on cache invalidation, so it uses its own id, simply to make sure that it is never mistaken for the default branch's cache by other syncs.

**Review by matburt — APPROVED** [review]

File: awx/main/models/projects.py
Comment (chrismeyersfsu) [inline awx/main/models/projects.py:472]: This isn't something you need to address, just FYI. Running a task now goes over NOTIFY/LISTEN interface. `NOTIFY` has a maximum payload size of 8,000. Apply that to this case and if the path string is longer than 8,000 characters, the cleanup will not run.

Here is a pattern when the payload is large.

https://github.com/ansible/awx/blob/d8f6ea3fe89eaa6aea375dfa0280b43c66899ae0/awx/api/generics.py#L987
https://github.com/ansible/awx/blob/d8f6ea3fe89eaa6aea375dfa0280b43c66899ae0/awx/main/tasks.py#L2938

**Review by chrismeyersfsu — APPROVED** [review]

File: awx/main/models/projects.py
Comment (AlanCoding) [reply awx/main/models/projects.py:472]: huh, I would think that could be a problem for the `handle_setting_changes` task, but it looks like the 8,000 limit may be specifically selected to be more than the character count for a listing of all setting keys, which comes to 6,845.

The project folder _could_ be anywhere if the user configured it in CTiT, but that's not a use case that concerns me a lot. The project name is in both of these paths, but that's capped at 512. It actually helps that we issue the task twice, instead of putting more data in the args.

**Pull request merged**
